cmd/util/ledger/migrations/atree_register_migration.go

```diff
@@ -0,0 +1,429 @@
+		delete(originalPayloads, id)
+
+		if len(value) == 0 {
+			// value was deleted
+			continue
+		}
+		ownerAddress, err := common.BytesToAddress([]byte(id.Owner))
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert owner address: %w", err)
+		}
+
+		if ownerAddress.Hex() != address.Hex() {
+			// something was changed that does not belong to this account. Log it.
+			m.log.Error().
+				Str("key", id.String()).
+				Str("owner_address", ownerAddress.Hex()).
+				Str("account", address.Hex()).
+				Msg("key is part of the change set, but is for a different account")
+
+			return nil, fmt.Errorf("register for a different account was produced during migration")
+		}
+
+		newPayloads = append(newPayloads, *ledger.NewPayload(util.RegisterIDToKey(id), value))
+	}
+
+	hasMissingKeys := false
+
+	// add all values that were not changed
+	for id, value := range originalPayloads {
+		if len(value) == 0 {
+			// this is strange, but we don't want to add empty values. Log it.
+			m.log.Warn().Msgf("empty value for key %s", id)
+			continue
+		}
+		if id.IsInternalState() {
+			// this is expected. Move it to the new payload
+			newPayloads = append(newPayloads, *ledger.NewPayload(util.RegisterIDToKey(id), value))
+			continue
+		}
+
+		isADomainKey := false
+		for _, domain := range domains {
+			if id.Key == domain {
+				isADomainKey = true
+				break
+			}
+		}
+		if isADomainKey {
+			// TODO: check if this is really expected
+			// this is expected. Move it to the new payload
+			newPayloads = append(newPayloads, *ledger.NewPayload(util.RegisterIDToKey(id), value))
+			continue
+		}
+
+		// something was not moved. Log it.
+		m.log.Debug().
+			Str("key", id.String()).
+			Str("account", address.Hex()).
+			Str("value", fmt.Sprintf("%x", value)).
+			Msg("Key was not migrated")
+		m.rw.Write(migrationError{
+			Address: address.Hex(),
+			Key:     id.String(),
+			Kind:    "not_migrated",
+			Msg:     fmt.Sprintf("%x", value),
+		})
+		hasMissingKeys = true
+	}
+
+	if hasMissingKeys && m.sampler.Sample(zerolog.InfoLevel) {
+		before := m.rwf.ReportWriter(fmt.Sprintf("account-before-%s", address.Hex()))
+		for _, p := range payloads {
+			before.Write(p)
+		}
+		before.Close()
+
+		after := m.rwf.ReportWriter(fmt.Sprintf("account-after-%s", address.Hex()))
+		for _, p := range newPayloads {
+			after.Write(p)
+		}
+		after.Close()
+	}
+	return newPayloads, nil
+}
+
+func (m *AtreeRegisterMigrator) checkStorageHealth(
+	address common.Address,
+	payloads []ledger.Payload,
+) error {
+	snapshot, err := util.NewPayloadSnapshot(payloads)
+	if err != nil {
+		return fmt.Errorf("failed to create payload snapshot: %w", err)
+	}
+
+	transactionState := state.NewTransactionState(snapshot, state.DefaultParameters())
+	accounts := environment.NewAccounts(transactionState)
+
+	accountsAtreeLedger := util.NewAccountsAtreeLedger(accounts)
+	storage := runtime.NewStorage(accountsAtreeLedger, util.NopMemoryGauge{})
```

Comment:
@janezpodhostnik 

I looked into `atree.CheckStorageHealth()` not detecting issues and it seems to be fine.  When tested outside this migration program, it was able to detect a problematic slab you provided on Friday.

Here, `storage.CheckHealth()` can fail to detect if slabs or maps are not loaded yet:
- `atree.CheckStorgeHealth()` checks loaded slabs
- `cadence.Storage` checks loaded storage maps

To fix failed detections on pre-migration slabs, try something like:
```
// load (but don't create) storage map in Cadence storage
.. = storage.GetStorageMap(address, "public", false)
.. = storage.GetStorageMap(address, "private", false)
.. = storage.GetStorageMap(address, "storage", false)

// load slabs in atree storage
for id := range payloadKeysToStorageIDs() {
   ... := storage.Retrieve(id)
}
```

Please let me know if this works.